### PR TITLE
KAFKA-1614: Partition log directory name and segments information exposed via JMX

### PR DESCRIPTION
**Pull request by xdralex**
This is an updated version of https://issues.apache.org/jira/browse/KAFKA-1614 patch which was made for 0.8.1.1 version. 

It makes partition log directory and single segments information exposed via JMX. This is useful to:
- monitor disks usage in a cluster and on single broker;
- calculate disk space taken by different topics;
- estimate space to be freed when segments are expired.

Yahoo's Kafka Manager uses it to show information about topics and partitions sizes:

topics
partitions

Our internal tool uses it to show how segments sizes were changing over time:

segments

@junrao could you please review when you have time?

### CODE REVIEW

File: core/src/main/scala/kafka/log/Log.scala [hunk 1]
```diff
@@ -292,6 +305,20 @@ class Log(val dir: File,
    */
   def numberOfSegments: Int = segments.size
 
+
+  /**
+   * Returns compiled information about segments
+   */
+  def segmentsInfo: util.ArrayList[String] = {
```
Comment (ijuma) [inline core/src/main/scala/kafka/log/Log.scala]: Given that we already expose the segments via `logSegments`, do we need to expose this method too? It may make sense to just implement it inside the `LogSegments` gauge.

File: core/src/main/scala/kafka/log/Log.scala [hunk 2]
```diff
@@ -292,6 +305,20 @@ class Log(val dir: File,
...
+    val list = new util.ArrayList[String]()
+    logSegments.foreach(
+      seg => list.add("baseOffset=%s, created=%s, logSize=%s, indexSize=%s".format(
+        seg.baseOffset, seg.created, seg.size, seg.index.sizeInBytes()
+      ))
+    )
+    list
```
Comment (ijuma) [inline core/src/main/scala/kafka/log/Log.scala]: This can be implemented more nicely by using `map`, eg

``` scala
import scala.collection.JavaConverters._
...
logSegments.toSeq.map(seg => s"baseOffset=${seg.baseOffset}, created=${seg.created}, logSize=${seg.size}, indexSize=$[seq.index.sizeInBytes}").asJava
```

Comment (xdralex) [inline core/src/main/scala/kafka/log/Log.scala]: I agree, but JConsole doesn't really show attribute values for Scala collections decorated as lists. Haven't digged deeper, but I suspect in this case it needs Scala classes in the classpath in order to show attributes correctly. It seems to me that the benefit of being able to use JConsole out of the box outweighs simplicity of the code in this case.

Example:

``` scala
import com.yammer.metrics.Metrics
import com.yammer.metrics.core.Gauge
import com.yammer.metrics.reporting.JmxReporter
import scala.collection.JavaConverters._

object Test
{
  def main(args: Array[String]) {
    val gaugeA = new Gauge[java.util.List[String]] {
      override def value(): java.util.List[String] = {
        val x = new java.util.LinkedList[String]()
        x.add("a")
        x.add("b")
        x.add("c")
        x
      }
    }

    val gaugeB = new Gauge[java.util.List[String]] {
      override def value(): java.util.List[String] = {
        val a = List("a", "b", "c")
        a.asJava
      }
    }

    Metrics.defaultRegistry().newGauge(this.getClass, "testA", gaugeA)
    Metrics.defaultRegistry().newGauge(this.getClass, "testB", gaugeB)

    val reporter = new JmxReporter(Metrics.defaultRegistry())
    reporter.start()

    Thread.sleep(1000 * 3600)
  }
}
```

testa
testb

Comment (ijuma) [inline core/src/main/scala/kafka/log/Log.scala]: That's good to know, thanks. It's worth adding a comment explaining that.

File: core/src/main/scala/kafka/log/Log.scala (hunk 1 above)
Comment (xdralex) [inline core/src/main/scala/kafka/log/Log.scala]: Sure thing, please take another look when possible.

File: core/src/main/scala/kafka/log/Log.scala [hunk 3]
```diff
@@ -133,6 +135,24 @@ class Log(val dir: File,
     },
     tags)
 
+  newGauge("LogSegments",
+    new Gauge[util.List[String]] {
+      def value = {
+        val list = logSegments.toSeq.map(
+          seg => s"baseOffset=${seg.baseOffset}, created=${seg.created}, logSize=${seg.size}, indexSize=${seg.index.sizeInBytes()}"
+        )
```
Comment (ijuma) [inline core/src/main/scala/kafka/log/Log.scala]: Thanks for addressing the feedback. Very minor point, but the following matches our style convention:

``` scala
logSegments.toSeq.map { seg =>
  s"baseOffset=${seg.baseOffset}, created=${seg.created}, logSize=${seg.size}, indexSize=${seg.index.sizeInBytes()}"
}
```

**Comment by ijuma** [comment]
This seems fine to me, let's see what @junrao says.

File: core/src/main/scala/kafka/log/Log.scala (hunk 3 above)
Comment (xdralex) [inline core/src/main/scala/kafka/log/Log.scala]: Sure, updated

**Comment by xvrl** [comment]
@ijuma any update on this?

**Comment by ijuma** [comment]
@xvrl, We need @junrao or another committer to review this.

**Comment by xvrl** [comment]
@ijuma if @junrao is not available, can you think of anyone else that might be able to review this?

**Comment by bender-the-greatest** [comment]
+1

File: core/src/main/scala/kafka/log/Log.scala [hunk 4]
```diff
@@ -133,6 +135,24 @@ class Log(val dir: File,
...
     tags)
 
+  newGauge("LogSegments",
+    new Gauge[util.List[String]] {
+      def value = {
+        val list = logSegments.toSeq.map { seg =>
+          s"baseOffset=${seg.baseOffset}, created=${seg.created}, logSize=${seg.size}, indexSize=${seg.index.sizeInBytes()}"
+        }
+        // Explicitly returning Java list to support JMX clients that don't have Scala runtime in the classpath
+        new util.ArrayList[String](list.asJava)
```
Comment (junrao) [inline core/src/main/scala/kafka/log/Log.scala:145]: The jmx metric value will use comma to delimit different segments. Since each segment also has comma separated fields, it makes it hard to parse the text value. Could we use a different delimiter within a segment?

**Comment by ijuma** [comment]
One concern I had was the cost of string concatenation for this metric. All other metrics are simple numbers. Is this a valid concern @junrao?

**Comment by junrao** [comment]
@ijuma : Yes, that's a good point. Instead of exposing this through jmx, another option is to add a new BrokerStatRequest and we can return all useful statistics in the response. This can potentially be useful for admin tools like partition reassignment.

**Comment by Antwnis** [comment]
+1 on a resolution on this (dated summer 2014) feature

**Comment by montana-ua** [comment]
This is patch for the kafka 0.10.x releases (existing patch does not work for 0.10.x)

```
diff --git a/core/src/main/scala/kafka/log/Log.scala b/core/src/main/scala/kafka/log/Log.scala
index a7549dc..6b7f199 100644
--- a/core/src/main/scala/kafka/log/Log.scala
+++ b/core/src/main/scala/kafka/log/Log.scala
@@ -23,6 +23,7 @@ import kafka.common._
 import kafka.metrics.KafkaMetricsGroup
 import kafka.server.{BrokerTopicStats, FetchDataInfo, LogOffsetMetadata}
 import java.io.{File, IOException}
+import java.util
 import java.util.concurrent.{ConcurrentNavigableMap, ConcurrentSkipListMap}
 import java.util.concurrent.atomic._
 import java.text.NumberFormat
@@ -133,6 +134,12 @@ class Log(val dir: File,
     },
     tags)

+    newGauge(name + "-" + "LogSegments",
+             new Gauge[util.ArrayList[String]] { def value = segmentsInfo })
+
+    newGauge(name + "-" + "Directory",
+             new Gauge[String] { def value = dir.getAbsolutePath })
+
   /** The name of this log */
   def name  = dir.getName()

@@ -293,6 +300,19 @@ class Log(val dir: File,
   def numberOfSegments: Int = segments.size

   /**
+   * Returns compiled information about segments
+   */
+  def segmentsInfo: util.ArrayList[String] = {
+    val list = new util.ArrayList[String]()
+    logSegments.foreach(
+      seg => list.add("baseOffset=%s, created=%s, logSize=%s, indexSize=%s".format(
+        seg.baseOffset, seg.created, seg.size, seg.index.sizeInBytes()
+      ))
+    )
+    list
+  }
+
+  /**
    * Close this log
    */
   def close() {
```

**Comment by simplesteph** [comment]
Can we see this merged ? +1

**Comment by schmichri** [comment]
+1

**Comment by murongruomeng** [comment]
@xdralex how about this function for kafka 10.0.2.1 or above?

**Comment by murongruomeng** [comment]
@montana-ua looks like it doesn't work for kafka 0.10.2.1

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/5827/
Test FAILed (JDK 8 and Scala 2.12).

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/5841/
Test FAILed (JDK 7 and Scala 2.11).

**Comment by xdralex** [comment]
@murongruomeng Sadly recently I don't have lots of time to maintain this patch outside of the master branch. Would be great if someone could take the effort to update it and get merged. Maybe @xvrl? :)

**Comment by everpcpc** [comment]
hi @xdralex , I've updated the patch, and tested on `1.1.0` with #5102 .
